src/zarr/abc/metadata.py

```diff
@@ -44,3 +46,23 @@ def from_dict(cls, data: dict[str, JSON]) -> Self:
         """
 
         return cls(**data)
+    
+    def __eq__(self, other: Any) -> bool:
```

Comment:
Overriding this means we lose the nice auto-generated dataclass `__eq__` method, which prints out like this:

```python
Matching attributes:
['shape',
 'data_type',
 'chunk_grid',
 'chunk_key_encoding',
 'codecs',
 'attributes',
 'dimension_names',
 'zarr_format',
 'node_type',
 'storage_transformers']
Differing attributes:
['fill_value']

Drill down into differing attribute fill_value:
  fill_value: np.float32(nan) != np.float32(nan)
```

But I don't see an easy way to keep that formatting and make it handle `fill_value` properly too...